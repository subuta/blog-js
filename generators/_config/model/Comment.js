export default {
  tableName: 'comments',
  required: [
    'text',
  ],

  properties: {
    id: {
      type: 'integer'
    },
    text: {
      'type': 'string'
    }
  },

  relations: {
    channel: {
      belongsTo: 'channels',
      join: {
        from: 'comments.channelId',
        to: 'channels.id'
      }
    },

    commentedBy: {
      belongsTo: 'users',
      join: {
        from: 'comments.commentedById',
        to: 'users.id'
      }
    },

    attachment: {
      hasOne: 'attachment',
      join: {
        from: 'comments.attachmentId',
        to: 'attachments.id'
      }
    },
  }
}
