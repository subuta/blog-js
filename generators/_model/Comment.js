export default {
  title: 'Comment',
  $id: 'http://sub-labo.com/schemas/comment.json',
  type: 'object',
  required: ['text'],
  properties: {
    text: {
      'type': 'string'
    }
  }
}
