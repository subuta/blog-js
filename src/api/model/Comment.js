import Model from './Model'
import _ from 'lodash'

export default class Comment extends Model {
  // SEE: https://github.com/Vincit/objection.js/issues/825
  $parseDatabaseJson(db) {
    // Remember to call the super class's implementation.
    const json = super.$parseDatabaseJson(db)

    return json
  }

  $formatDatabaseJson(json) {
    return super.$formatDatabaseJson(json)
  }

  static register = (models) => {
    // then define relationMappings.
    Comment.relationMappings = {
      channel: {
        modelClass: models.Channel,
        relation: Model.BelongsToOneRelation,
        join: {from: 'comments.channelId', to: 'channels.id'}
      },
      commentedBy: {
        modelClass: models.User,
        relation: Model.BelongsToOneRelation,
        join: {from: 'comments.commentedById', to: 'users.id'}
      },
      attachment: {
        modelClass: models.Attachment,
        relation: Model.HasOneRelation,
        join: {from: 'comments.attachmentId', to: 'attachments.id'}
      }
    }
  }

  static tableName = 'comments'

  static jsonSchema = {
    title: 'Comment',
    $id: 'http://sub-labo.com/schemas/comment.json',
    type: 'object',
    required: ['text'],
    properties: {id: {type: 'integer'}, text: {type: 'string'}}
  }
}
