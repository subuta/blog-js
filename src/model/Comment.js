import sequelize from 'src/utils/sequelize'

const Sequelize = sequelize.constructor

const Comment = sequelize.define('comment', {
  id: {
    type: Sequelize.INTEGER,
    allowNull: false,
    primaryKey: true,
    autoIncrement: true
  },
  text: {
    type: Sequelize.TEXT
  },
  createdAt: {
    type: Sequelize.DATE,
    allowNull: false,
    defaultValue: Sequelize.NOW
  },
  updatedAt: {
    type: Sequelize.DATE,
    allowNull: false,
    defaultValue: Sequelize.NOW
  }
})

export default Comment
