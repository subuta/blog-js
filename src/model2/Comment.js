import Model from './Model'

export const register = (models) => {
  Comment.relationMappings = {}
}

export default class Comment extends Model {
  static tableName = 'comments'

  static jsonSchema = {
    title: 'Comment',
    $id: 'http://sub-labo.com/schemas/comment.json',
    type: 'object',
    required: ['text'],
    properties: {text: {type: 'string'}}
  }
}
