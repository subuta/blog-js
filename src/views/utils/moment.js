import 'moment/locale/en-gb'
import moment from 'moment'

export default moment
