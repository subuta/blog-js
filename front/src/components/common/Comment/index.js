import React from 'react'
import moment from 'src/utils/moment'
import _ from 'lodash'

import classes from './style'

import Avatar from 'src/components/common/Avatar'

export default ({comment}) => {
  const {
    id,
    text,
    attachment,
    commentedBy
  } = comment

  const createdAt = moment(comment.createdAt).format('A HH:mm')
  const avatar = _.get(commentedBy, 'avatar')
  const nickname = _.get(commentedBy, 'nickname')

  return (
    <div className={classes.CommentWrapper}>
      <Avatar avatar={avatar} nickname={nickname} />

      <div className={classes.Comment}>
        <div className={classes.Nickname}>
          <span>{nickname || 'Anonymous'}</span>

          <small className={classes.CommentedAt}>{createdAt}</small>
        </div>

        {attachment && (
          <img src={attachment.url} alt={attachment.name} />
        )}
        <p>{text}</p>
      </div>
    </div>
  )
}
