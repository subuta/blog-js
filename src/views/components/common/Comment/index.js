import React from 'react'
import moment from 'src/views/utils/moment'
import _ from 'lodash'

import MdCreateIcon from 'react-icons/lib/md/create'
import MdDeleteIcon from 'react-icons/lib/md/delete'

import MdInsertEmoticonIcon from 'react-icons/lib/md/insert-emoticon'
import MarkdownContent from 'src/views/components/common/MarkdownContent'
import Reactions from 'src/views/components/common/Reactions'

import withStyles from './style'

import {
  compose,
  withState,
  lifecycle,
  withPropsOnChange
} from 'recompose'

import Avatar from 'src/views/components/common/Avatar'
import Tooltip from 'src/views/components/common/Tooltip'

import { toHtml } from 'src/views/utils/markdown'

const isBrowser = typeof window !== 'undefined'

const enhance = compose(
  withStyles,
  withState('isHover', 'setIsHover', false),
  lifecycle({
    componentDidMount () {
      const {
        attachment,
        onLoad = _.noop
      } = this.props

      // Call onLoad at this hook for no-attachment comment.
      if (!attachment) {
        onLoad()
      }
    }
  })
)

export default enhance((props) => {
  const {
    className,
    comment,
    styles,
    style,
    isHover,
    setIsHover,
    isAuthenticated,
    currentUser,
    onLoad = _.noop, // Will be called on content loaded.
    onEdit = _.noop,
    onDelete = _.noop,
    onAddReaction = _.noop,
    onRemoveReaction = _.noop
  } = props

  const {
    id,
    text,
    attachment,
    commentedBy,
    reactions
  } = comment

  const createdAt = moment(comment.created_at).format('A HH:mm')
  const avatar = _.get(commentedBy, 'avatar')
  const nickname = _.get(commentedBy, 'nickname')

  let commentWrapperClass = styles.CommentWrapper
  if (isHover) {
    commentWrapperClass += ' is-hovered'
  }

  if (className) {
    commentWrapperClass += ` ${className}`
  }

  return (
    <div
      className={commentWrapperClass}
      onMouseEnter={() => setIsHover(true)}
      onMouseLeave={() => setIsHover(false)}
      style={style}
    >
      <Avatar avatar={avatar} nickname={nickname}/>

      <div className={styles.Comment}>
        <div className={styles.Nickname}>
          <span>{nickname || 'Anonymous'}</span>

          <small className={styles.CommentedAt}>{createdAt}</small>
        </div>

        {attachment && (
          <img
            src={attachment.imageUrl}
            alt={attachment.name}
            onLoad={onLoad}
          />
        )}

        <MarkdownContent
          className="text"
          html={toHtml(text)}
        />

        <Reactions
          reactions={comment.reactions}
          onClose={() => setIsHover(false)}
          onAddReaction={(emoji) => onAddReaction(comment, emoji)}
          onRemoveReaction={(emoji) => onRemoveReaction(comment, emoji)}
          disabled={!isAuthenticated}
          currentUser={currentUser}
          mini
        />
      </div>

      {isHover && (
        <span className={styles.Actions}>
          <Reactions
            className={styles.Action}
            reactions={[]}
            onClose={() => setIsHover(false)}
            onAddReaction={(emoji) => onAddReaction(comment, emoji)}
            disabled={!isAuthenticated}
            currentUser={currentUser}
            embedded
          />

          <Tooltip
            title="Edit this comment"
            placement="top"
            size="small"
          >
            <span
              className={styles.Action}
              onClick={onEdit}
            >
              <MdCreateIcon/>
            </span>
          </Tooltip>

          <Tooltip
            title="Delete this comment"
            placement="top"
            size="small"
          >
            <span
              className={styles.Action}
              onClick={onDelete}
            >
            <MdDeleteIcon/>
          </span>
          </Tooltip>
        </span>
      )}
    </div>
  )
})
