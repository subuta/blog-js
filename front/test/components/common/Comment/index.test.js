/* global describe, it */
import React from 'react'
import { create } from 'react-test-renderer'
import moment from 'moment'

import Comment from 'src/components/common/Comment'

describe('Comment', () => {
  it('should be rendered correctly', () => {
    const tree = create(
      <Comment
        comment={{
          text: 'hoge',
          attachment: {
            url: 'http://lorempixel.com/200/200/cats/',
            name: 'hoge.png'
          },
          commentedBy: {
            avatar: 'http://lorempixel.com/200/200/cats/',
            nickname: 'Taro Yamada'
          },
          createdAt: moment('2017/12/01 18:00:00', 'YYYY/MM/DD HH:mm:ss')
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should be rendered correctly without commentedBy', () => {
    const tree = create(
      <Comment
        comment={{
          text: 'hoge',
          attachment: {
            url: 'http://lorempixel.com/200/200/cats/',
            name: 'hoge.png'
          },
          createdAt: moment('2017/12/01 18:00:00', 'YYYY/MM/DD HH:mm:ss')
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
