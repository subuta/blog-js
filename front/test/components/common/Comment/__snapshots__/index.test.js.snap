// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should be rendered correctly 1`] = `
<div
  className="f12b9rze"
>
  <div
    className="f6deexk"
  >
    <img
      alt="Taro Yamada"
      src="http://lorempixel.com/200/200/cats/"
    />
  </div>
  <div
    className="f8xzse3"
  >
    <div
      className="f1sfq9p6"
    >
      <span>
        Taro Yamada
      </span>
      <small
        className="fq0fzr3"
      >
        午後 18:00
      </small>
    </div>
    <img
      alt="hoge.png"
      src="http://lorempixel.com/200/200/cats/"
    />
    <p>
      hoge
    </p>
  </div>
</div>
`;

exports[`Comment should be rendered correctly without commentedBy 1`] = `
<div
  className="f12b9rze"
>
  <div
    className="f6deexk"
  >
    <div
      className="f1y5k9zg"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m20 33.4c7.3 0 13.4-6.1 13.4-13.4 0-1.3-0.3-2.6-0.6-3.7-1.2 0.3-2.4 0.3-3.7 0.3-5.7 0-10.6-2.7-13.6-7-1.7 4-4.9 7.3-8.8 9-0.1 0.5-0.1 0.9-0.1 1.4 0 7.3 6.1 13.4 13.4 13.4z m0-30c9.2 0 16.6 7.4 16.6 16.6s-7.4 16.6-16.6 16.6-16.6-7.4-16.6-16.6 7.4-16.6 16.6-16.6z m5 16.2c1.2 0 2.1 0.9 2.1 2s-0.9 2.2-2.1 2.2-2.1-1-2.1-2.2 0.9-2 2.1-2z m-10 0c1.2 0 2.1 0.9 2.1 2s-0.9 2.2-2.1 2.2-2.1-1-2.1-2.2 0.9-2 2.1-2z"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="f8xzse3"
  >
    <div
      className="f1sfq9p6"
    >
      <span>
        Anonymous
      </span>
      <small
        className="fq0fzr3"
      >
        午後 18:00
      </small>
    </div>
    <img
      alt="hoge.png"
      src="http://lorempixel.com/200/200/cats/"
    />
    <p>
      hoge
    </p>
  </div>
</div>
`;
